include/swift/SIL/OwnershipLiveness.h

```diff
@@ -0,0 +1,266 @@
+//===--- OwnershipLiveness.h ---------------------------------*- C++ -*----===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// Terminology:
+///
+/// Linear lifetime: All paths through the value's definition to the
+/// function exit pass through exactly one lifetime-ending operation.
+///
+/// Complete OSSA: All owned values and borrow introducers have linear lifetime.
+///
+///
+/// Linear SSA liveness:
```

Comment:
> what is "liveness" (compared to "lifetime").

Good question. It's not that important to distinguish, but I find it useful.

A lifetime means different things depending on context:
- Refcounted objects have a dynamic lifetime from allocation to deinit
- In lexical SIL, variable declaration have a static lifetime
- In ownership SIL, every owned value has a lifetime

In ownership SIL, rewriting or optimizing a lifetime requires computing liveness.

Liveness is the data flow indicating where values are used. Liveness can be interpreted in different ways depending on the context. We often need to extend liveness incrementally as we discover things that should be "covered" by the value's lifetime. We may conditionally extend it across access scopes, debug information, deinit barriers, etc.

Once all things are considered, you get back a liveness boundary (essentially an InstructionRange), then use that to do perform some query (check if certain things are "contained" by the boundary), or run some diagnostics. Sometimes, you want to keep liveness around after the check, because you might further extend it by adding more uses (after replace-all-uses). Note that `isWithinBoundary` is a lot more efficient though when you still have the liveness information, vs. just the boundary.

These interfaces are called "liveness" because they compute liveness and store the data flow result internally. The choice of algorithm depends of the kind of lifetime you're dealing with. For example, is the lifetime already valid, or did you just transform the SIL in a way that made it invalid.